src/hypertable.c

```diff
@@ -897,14 +905,17 @@ hypertable_create(PG_FUNCTION_ARGS)
 	NameData	schema_name,
 				table_name,
 				default_associated_schema_name;
+	Relation	rel;
 
 	/*
 	 * Serialize hypertable creation to avoid having multiple transactions
 	 * creating the same hypertable simultaneously. The lock should conflict
 	 * with itself and RowExclusive, to prevent simultaneous inserts on the
-	 * table. Hold until transaction end.
+	 * table. Also this should conflict with SELECT as it potentially moves
+	 * the data into other tables due to migration. Hold until transaction end
+	 * unless you hit if not exists case.
 	 */
-	LockRelationOid(table_relid, ShareRowExclusiveLock);
+	rel = heap_open(table_relid, AccessExclusiveLock);
```

Comment:
Should we wait with opening the heap until after hypertable check and just lock here like before? Otherwise you might open the heap for no reason if the hypertable already exists.

Reply:
Also, if  if-not-exists is not set and this is already a hypertable, is it OK to error out without closing the heap? Can't we  heap-close and unlock in either case?